src/D3MHub.sol

```diff
@@ -221,121 +219,118 @@ contract D3MHub {
-        // That's why it converts normalized debt (art) to Vat DAI generated with a simple RAY multiplication or division
-        // This module will have an unintended behaviour if rate is changed to some other value.
-
-        EndLike _end;
-        uint256 daiDebt;
-        if (mode == Mode.NORMAL) {
-            // Normal mode or module just caged (no culled)
-            // debt is obtained from CDP art
-            (, daiDebt) = vat.urns(ilk, address(_pool));
-        } else if (mode == Mode.D3M_CULLED) {
-            // Module shutdown and culled
-            // debt is obtained from free collateral owned by the pool contract
-            // We rebalance the CDP after grabbing in `cull` so the gems represents
-            // the debt at time of cull
-            daiDebt = vat.gem(ilk, address(_pool));
-        } else if (mode == Mode.MCD_CAGED) {
-            // MCD caged
-            // debt is obtained from free collateral owned by the End module
-            _end = end;
-            _end.skim(ilk, address(_pool));
-            daiDebt = vat.gem(ilk, address(_end));
-        } else revert("D3MHub/unknown-mode");
-
-        uint256 availableAssets = _pool.maxWithdraw();
-
-        // Unwind amount is limited by how much:
-        // - max reduction desired
-        // - assets available
-        // - dai debt tracked in vat (CDP or free)
-        uint256 amount = _min(
-                            _min(
-                                supplyReduction,
-                                availableAssets
-                            ),
-                            daiDebt
-                        );
-        require(amount <= MAXINT256, "D3MHub/overflow");
+    function _unwind(bytes32 ilk, ID3MPool _pool, uint256 amount) internal {
+        if (amount > 0) {
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
+            vat.frob(ilk, address(_pool), address(_pool), address(this), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(_pool), -int256(amount));
+        }
+        emit Unwind(ilk, amount);
+    }
 
-        // Determine the amount of fees to bring back
-        uint256 fees = 0;
-        if (assetBalance > daiDebt) {
-            unchecked {
-                fees = assetBalance - daiDebt;
-            }
+    function _unwind(bytes32 ilk, ID3MPool _pool, address urn) internal {
+        uint256 amount = _pool.maxWithdraw();
+        if (amount > 0) {
+            uint256 toSlip = _min(vat.gem(ilk, urn), amount);
+            require(toSlip <= MAXINT256, "D3MHub/overflow");
+            vat.slip(ilk, urn, -int256(toSlip));
 
-            if (amount + fees > availableAssets) {
-                unchecked {
-                    fees = availableAssets - amount;
-                }
-            }
+            _pool.withdraw(amount);
+            daiJoin.join(address(this), amount);
```

Comment:
Just to confirm my understanding here...we intentionally leave `ink` in the pool's urn b/c there is no reason to `grab` and `slip` it away?

Reply:
@kmbarry1 can you explain why you expect `ink` to be positive?
This version of `_unwind` (with the address as 3rd parameter) can only be called if the `ilk` is culled (in which case `ink` was grabbed) or after skimming (in which case `ink` is 0 since `ink == art`).

Reply:
The only exception is when there is permissionless DAI repaid right after ES. Otherwise yeah, I'm a bit lost the case Kurt might be asking.

Reply:
Yeah, believe I was thinking about the case Gonza highlights. Worst case I guess some lending pool shares get stuck in the D3M? This isn't a normal Vault where the owner can claim their collateral with a `free` call.

Reply:
Yeah the only case where `ink` might be higher than `art` is if there was a permissionless repayment of DAI right before ES (not after as I said in the prev comment). As there wasn't time to recall `exec` to fix it. In that case that difference of `ink` will be lost.

Then the other thing that might be lost is the different between the `ink` and the real amount of assets accrued in the pool. That won't be updated after ES in `exec`.
So if before `end.wait` finishes, there is liquidity, all of it will be recovered, but otherwise only `ink` amount will be shared to DAI holders.